Add TypeSpec for Azure.Communication.Messages service

glorialimicrosoft opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [x] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

glorialimicrosoft committed:
Add GA swagger

glorialimicrosoft committed:
add typespec

glorialimicrosoft committed:
generated swagger by TypeSpec

glorialimicrosoft committed:
add repeatability headers

glorialimicrosoft committed:
Update client.tsp

glorialimicrosoft committed:
try adding internal to the methods

glorialimicrosoft committed:
mark response public

glorialimicrosoft committed:
make clients internal for all languages

glorialimicrosoft committed:
mark more as public

glorialimicrosoft committed:
fix naming - StreamClient

glorialimicrosoft committed:
remove StreamClient, move into NotificationMessagesClient

glorialimicrosoft committed:
fix typo - application/octet-stream

glorialimicrosoft committed:
change body to content

glorialimicrosoft committed:
renaming

glorialimicrosoft committed:
some changes to model access

glorialimicrosoft committed:
change the access decorator to csharp only

glorialimicrosoft committed:
experiment with polymorphism

glorialimicrosoft committed:
Update client.tsp

glorialimicrosoft committed:
add @access(Access.public, "csharp"), fix WhatsAppMessageTemplateValueSubType

glorialimicrosoft committed:
Update models.tsp

glorialimicrosoft committed:
Update models.tsp

glorialimicrosoft committed:
add name as a field to MessageTemplateValue

glorialimicrosoft committed:
add @visibility("create")

glorialimicrosoft committed:
try @visibility("read", "query", "create", "update", "delete")

glorialimicrosoft committed:
try Array<MessageTemplateValue>

glorialimicrosoft committed:
change client

glorialimicrosoft committed:
make MessageTemplateResponse and WhatsAppMessageTemplateResponse internal

glorialimicrosoft committed:
Update CommunicationServicesMessages.json

glorialimicrosoft committed:
more discriminators

glorialimicrosoft committed:
Update models.tsp

glorialimicrosoft committed:
Update models.tsp

glorialimicrosoft committed:
Update models.tsp

glorialimicrosoft committed:
address comment for adding SupportsClientRequestId

"ServiceTraits should also includeSupportsClientRequestId because that header is mandatory for all Azure services and will get auto-populated by the SDK."

DominikMe committed:
fix tsp compile warnings

DominikMe committed:
generate csharp for test purposes

DominikMe committed:
remove csharp client tweaks

DominikMe committed:
fix client hierarchy

DominikMe committed:
add auth decoration

DominikMe committed:
avoid response and request in model names

DominikMe committed:
avoid single noun names

DominikMe committed:
revert string literal to enum to control naming in SDK

DominikMe committed:
change button to buttons

DominikMe committed:
make enums public and rename client methods

glorialimicrosoft committed:
change operation name from sendMessage to send

DominikMe commented on specification/communication/Communication.Messages/client.tsp:
```diff
@@ -0,0 +1,28 @@
+})
+interface NotificationMessagesClient extends AzureCommunicationMessagesService.NotificationMessagesOperations
+{
+  @access(Access.internal)
```
Do we need to hide it only for C# where we replace it with a Stream type or also for other languages?

glorialimicrosoft replied:
So far we only received recommendation to use Stream in .NET and not other languages.
I've raised a question in the group chat with Azure board.

DominikMe commented on specification/communication/Communication.Messages/examples/2024-02-01/NotificationMessages_Send.json:
```diff
@@ -0,0 +1,29 @@
+      "to": [
+        "14250000000"
+      ],
+      "type": "text",
```
I think this needs to be `kind` now

glorialimicrosoft replied:
updated

DominikMe commented on specification/communication/Communication.Messages/examples/2024-02-01/Stream_GetStream.json:
```diff
@@ -0,0 +1,15 @@
+  },
+  "responses": {
+    "200": {
+      "body": {}
```
Can we change this to some example data? Maybe also add the octet-stream response header

DominikMe commented on specification/communication/Communication.Messages/examples/2024-02-01/Template_List.json:
```diff
@@ -0,0 +1,37 @@
+            "whatsApp": {
+              "content": [
+                {
+                  "type": "BODY",
```
This example doesn't match the model anymore, it seems?

glorialimicrosoft replied:
updated

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+@doc("A request to send a text notification.")
+model TextNotificationContent extends NotificationContent {
+  @doc("Message notification type is text.")
+  kind: "text";
```
Should we extract this literal value out to a new `enum NotificationContentType`?

glorialimicrosoft replied:
extracted to enum CommunicationMessageType

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+@doc("The message template's text value information.")
+model MessageTemplateText extends MessageTemplateValue {
+  @doc("Message parameter type is text.")
+  kind: "text";
```
Should we extract the literal out to an `enum MessageTemplateValueType`? Is there a difference in the generated SDK when we use a string literal vs an enum or do the APIs look the same?

DominikMe replied:
was extracted

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+  address?: string;
+
+  @doc("The latitude of the location.")
+  latitude?: float64;
```
The other two prop docs mention they're optional, so should latitude and longitude be required?

DominikMe replied:
made required

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+@doc("The template bindings for WhatsApp")
+model WhatsAppMessageTemplateBindings extends MessageTemplateBindings {
+  @doc("MessageTemplateBindings is whatsApp")
+  kind: "whatsApp";
```
Same comment as elsewhere, let's clarify if this should be extracted to an `MessageTemplateBindingsType` enum.

DominikMe replied:
was extracted

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+@doc("The template bindings component for WhatsApp")
+model WhatsAppMessageTemplateBindingsComponent {
+  @doc("The name of the referenced item in the template values.")
+  refValue: string;
```
Can we look up why we wrapped this once more, I'm unclear about the exact reason.

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+}
+
+@doc("The template bindings component for WhatsApp")
+model WhatsAppMessageTemplateBindingsComponent {
```
If we get feedback to flatten this in the SDK, we can try the @flattenProperties decorator, but maybe it's not supported in all languages yet, and maybe flattening isn't really a big benefit and there's value in being very close to the REST API.

weidongxu-microsoft replied:
I think almost all SDK arch wants to drop the flattenProperty (that was in autorest) for data-plane.

glorialimicrosoft replied:
Will keep it this way.

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+  subType?: WhatsAppMessageButtonSubType;
+
+  @doc("The name of the referenced item in the template values.")
+  refValue: string;
```
Ah, maybe we wrapped with WhatsAppMessageTemplateBindingsComponent to be consistent with WhatsAppMessageTemplateBindingsButton here?

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+This is the payload returned from WhatsApp
+API.
+""")
+  content?: unknown;
```
I think TypeSpec gives a warning for using unknown, can we add a suppression with a reason here? See JobRouter's tsp.
(If we know that this can be of `Record` type we can use that instead, but I assume it's not guaranteed to be an object?)

glorialimicrosoft replied:
added a suppression

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+  content?: unknown;
+
+  @doc("Message template response type is whatsApp.")
+  kind: "whatsApp";
```
Same question as elsewhere, should we extract an enum `MessageTemplateItemType`?

glorialimicrosoft replied:
extracted to enum CommunicationMessagesChannel

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,322 @@
+
+@resource("messages/streams")
+@doc("A data stream.")
+@access(Access.internal)
```
Should this be internal for all languages or only C#? Depends on how we're gonna represent this in other languages.

DominikMe commented on specification/communication/Communication.Messages/routes.tsp:
```diff
@@ -0,0 +1,37 @@
+
+interface StreamOperations {
+  @tag("Stream")
+  @doc("Download the Media payload from a User to Business message.")
```
Sometimes we use "media", and sometimes "stream" to describe the same thing. Can we look up from prior reviews why we named this /streams, and double-check that we're consistent in our terminology?

glorialimicrosoft replied:
The operation namespace is Stream, and the operation is called downloadMedia.

glorialimicrosoft replied:
Discussed in the review meeting, will keep it this way for now, until we can come up with a better name.

DominikMe commented on specification/communication/data-plane/Messages/stable/2024-02-01/examples/NotificationMessages_Send.json:
```diff
@@ -0,0 +1,29 @@
+      "to": [
+        "14250000000"
+      ],
+      "type": "text",
```
Should this be `kind`?

DominikMe replied:
Ah, this is generated, same comments as under Communication.Messages :-)

glorialimicrosoft replied:
updated and regenerated

DominikMe reviewed (commented):
This looks really clean! I've added some comments, which are mostly questions.

arifibrahim4 committed:
JS and Java SDK related typespec changes (#27416)

* updated js emitter output

* JS discriminator model issue  fix

* JS generation fix

* reverting the discriminator changes

* java emitter changes

* added key credential auth

* address some PR comments (#27435)

* renaming

* make latitude and longitude requried

* rename MessageTemplateType to MessageTemplateBindingsType

* add suppression for unknown type

* extract an enum CommunicationMessagesChannel for MessageTemplateItem

* fix examples

* fix examples and generate swagger

---------

Co-authored-by: glorialimicrosoft <67171398+glorialimicrosoft@users.noreply.github.com>

glorialimicrosoft committed:
remove outdated generated examples

glorialimicrosoft committed:
renaming

glorialimicrosoft committed:
use @projectedName("csharp", "uri")  to rename url to uri for csharp and make downloadMedia internal for csharp

DominikMe commented:
@glorialimicrosoft Can you please work through the CI issues so that the PR passes the checks? Thanks

glorialimicrosoft committed:
fix example, fix typo

glorialimicrosoft committed:
fix XmsEnumValidation error on build pipeline

glorialimicrosoft committed:
regenerate swagger with TypeSpec

glorialimicrosoft committed:
fix example

glorialimicrosoft committed:
fix nextLink in example

glorialimicrosoft committed:
fix nextLink model validation

TypeSpec automatically generates nextLink as an Uri type, and in our service we use relative path for nextLink, so validation is failing, setting it to null to bypass this validation

glorialimicrosoft committed:
run npx tsv specification/communication/Communication.Messages

glorialimicrosoft committed:
run tsp compile

glorialimicrosoft committed:
try something

glorialimicrosoft committed:
merge main into feature branch then rerun npx tsv specification/communication/Communication.Messages

glorialimicrosoft committed:
make communicationservicesmessages all lower case

glorialimicrosoft committed:
run npx tsv specification/communication/Communication.Messages again

glorialimicrosoft committed:
Update tspconfig.yaml

glorialimicrosoft committed:
fix Swagger Avocado

glorialimicrosoft committed:
fix Swagger Avocado

glorialimicrosoft committed:
add suppressions for LroExtension

glorialimicrosoft committed:
remove customization-class setting in tspconfig.yaml that is causing the customization task to fail.

glorialimicrosoft committed:
Address comment: rename getTemplates  in routes.tsp to listTemplates

glorialimicrosoft commented on specification/communication/Communication.Messages/routes.tsp:
```diff
@@ -0,0 +1,41 @@
+interface TemplateOperations {
+  @tag("Template")
+  @doc("List all templates for given ACS channel")
+  getTemplates is Operations.ResourceList<MessageTemplateItem>;
```
addressed the comment

arifibrahim4 commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+@doc("The template bindings component button for WhatsApp")
+model WhatsAppMessageTemplateBindingsButton {
+  @doc("The WhatsApp button sub type")
+  subType?: WhatsAppMessageButtonSubType;
```
`subType` is required field if we don't send it, causes 500 when mapping to internal object

glorialimicrosoft replied:
changed

glorialimicrosoft committed:
remove `?` from subType as it is required for WhatsAppMessageTemplateBindingsButton

mikekistler commented:
Please update the PR description to indicate if this API version is private or public preview:

<img width="258" alt="image" src="https://github.com/Azure/azure-rest-api-specs/assets/85643503/db4f3fd4-000a-40ef-9e5b-06ffca636157">

mikekistler commented on specification/communication/data-plane/Messages/stable/2024-02-01/communicationservicesmessages.json:
```diff
@@ -0,0 +1,1051 @@
+      },
+      "tokenUrl": "https://login.microsoftonline.com/common/oauth2/token"
+    },
+    "ApiKeyAuth": {
+      "type": "apiKey",
+      "name": "Authorization",
+      "in": "header"
+    }
```
Security definition should have a description.

glorialimicrosoft replied:
This swagger is auto gnerated by TypeSpec

mikekistler commented on specification/communication/data-plane/Messages/stable/2024-02-01/communicationservicesmessages.json:
```diff
@@ -0,0 +1,1051 @@
+      "description": "The WhatsApp-specific template response contract",
+      "properties": {
+        "content": {
+          "description": "WhatsApp platform's template content\nThis is the payload returned from WhatsApp\nAPI."
```
Recommend removing the multiple "\n"s in this description.

glorialimicrosoft replied:
removed by removing multiple-line docs from TypeSpec

mikekistler reviewed (approved):
Looks good. 👍 

Just a few very minor comments.

johanste commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+}
+
+@doc("The type of the message template.")
+enum MessageTemplateBindingsType {
```
Do we expect the number of bindings types to grow over time?

glorialimicrosoft replied:
yes it will grow as we introduce more channels

johanste commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+import "@typespec/http";
+import "@typespec/versioning";
+import "@azure-tools/typespec-azure-core";
+import "@azure-tools/typespec-client-generator-core";
```
Why do we need this import? We should constrain this to the client.tsp file and use augment (e.g. `@@access(CommunicationMessageType, Access.public)`

glorialimicrosoft replied:
Moved them out of models.tsp, into client.tsp

heaths commented on specification/communication/Communication.Messages/main.tsp:
```diff
@@ -0,0 +1,32 @@
+)
+@versioned(Versions)
+@doc("Azure Communication Messages Service")
+namespace AzureCommunicationMessagesService;
```
Should be formatted like Azure.Communication.MessagesService or something like that.

glorialimicrosoft replied:
fixed

heaths commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+@versioned(AzureCommunicationMessagesService.Versions)
+namespace AzureCommunicationMessagesService;
+
+@access(Access.public)
```
This all should default to public, so all these declarations are extraneous, I believe.

glorialimicrosoft replied:
removed

heaths commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+}
+
+@doc("The type of the message template.")
+enum MessageTemplateBindingsType {
```
We prefer "kind" over "type" for discriminators.

heaths replied:
Same for any enums or model property names throughout these specs.

glorialimicrosoft replied:
Changed to `Kind`

weidongxu-microsoft commented on specification/communication/Communication.Messages/.gitignore:
Any precedence in spec repo that we add `.gitignore` file into subfolder?

glorialimicrosoft replied:
We do not have a .gitignore in the parent folder, thus it is needed here. Resolving for now, this would be something to be considered in the future to move this out of project subfolder.

heaths commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+  @doc("Name of the template.")
+  name: string;
+
+  @doc("The codes for the supported languages for templates.")
```
In what format? Other services specify the spec e.g. BCP47 e.g., "en", "en-US", etc. Specify in you doc string.

glorialimicrosoft replied:
Added the format and more description

heaths commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+  name: string;
+
+  @doc("The template's language.")
+  language: string;
```
Same here: document the format. Do so for any language specification throughout.

glorialimicrosoft replied:
Documented

heaths reviewed (changes requested):
Given this is for a GA, there are a few changes I'd like to see first. Has this previewed, though?

weidongxu-microsoft commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,370 @@
+  @doc("MessageTemplateBindings is whatsApp")
+  kind: MessageTemplateBindingsType.whatsApp;
+
+  @doc("The header template bindings")
+  header?: WhatsAppMessageTemplateBindingsComponent[];
+
+  @doc("The body template bindings")
+  body?: WhatsAppMessageTemplateBindingsComponent[];
+
+  @doc("The footer template bindings")
+  footer?: WhatsAppMessageTemplateBindingsComponent[];
+
+  @doc("The button template bindings")
+  buttons?: WhatsAppMessageTemplateBindingsButton[];
```
nit, all of them is an array, why only `buttons` got plural?

glorialimicrosoft replied:
Discussed in the review meeting.

glorialimicrosoft committed:
add more examples

glorialimicrosoft committed:
format namespace as Azure.Communication.MessagesService

glorialimicrosoft committed:
remove unnecessary @access(Access.public)

glorialimicrosoft committed:
rename Type to Kind

glorialimicrosoft committed:
add description for language in template

glorialimicrosoft committed:
Update communicationservicesmessages.json

glorialimicrosoft committed:
fix CI failure, rename download to get

glorialimicrosoft committed:
remove multiple-line docs from TypeSpec

heaths commented:
Opened Azure/azure-sdk-tools#7593 for the APIView emitter crash.

glorialimicrosoft committed:
add maxpagesize for list template operation

glorialimicrosoft committed:
move @access out of models.tsp into client.tsp

DominikMe commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,364 @@
+alias ListTemplatesQueryParams = ListQueryParametersTrait<{
+  @doc("Number of objects to return per page.")
+  @query
+  maxpagesize?: int32;
```
@glorialimicrosoft Can we try using the https://azure.github.io/typespec-azure/docs/libraries/azure-core/reference/data-types#Azure.Core.MaxPageSizeQueryParameter type instead of defining your own?

glorialimicrosoft replied:
Actually this is not needed here, ListTemplatesQueryParams already takes care of it. Removed.

tjprescott commented on specification/communication/Communication.Messages/models.tsp:
```diff
@@ -0,0 +1,364 @@
+
+@doc("The type of message. Supports text, image, template.")
+enum CommunicationMessageKind {
+  @doc("Text message type.")
```
Generally, we recommend using `/** docstring */` syntax over `@doc` because you will get better IDE syntax highlighting.

glorialimicrosoft replied:
Thanks for the suggestion, this is how it is being written in other services too: https://github.com/Azure/azure-rest-api-specs/blob/6cecaf43ab9c1104e683aea3b185b88f81be2750/specification/communication/Communication.JobRouter/models.tsp
Is there a strong preference on changing to `/** docstring */`?

tjprescott replied:
It's mostly for your benefit. I don't think the Stewardship board or archboard reviewers even see these (by design).

heaths replied:
I often turn them on when reviewing to help explain what I'm reviewing. Not required to change, but I will admit that `/** */` comments syntax highlight better. Still, I don't see a need to change this. Maybe next service.

glorialimicrosoft committed:
remove the unneeded maxpagesize in models, ListTemplatesQueryParams already takes care of it

glorialimicrosoft committed:
Fix some doc description

glorialimicrosoft committed:
rerun  npx tsv specification/communication/Communication.Messages

glorialimicrosoft committed:
change channel id for ListTemplates to uuid

tjprescott commented on specification/communication/data-plane/Messages/readme.md:
````diff
@@ -40,13 +51,7 @@ title:
-See configuration in [readme.csharp.md](./readme.csharp.md)
+``` yaml
+suppressions:
+  - code: LroExtension
````
What is this suppression for?

glorialimicrosoft replied:
It is here to suppress "Operations with a 202 response must specify x-ms-long-running-operation: true." It is advised by Laurent Mazuel to add this suppression. It is safe to ignore as the service will send the long-running response through EG events.

glorialimicrosoft commented on specification/communication/data-plane/Messages/stable/2024-02-01/communicationservicesmessages.json:
```diff
@@ -0,0 +1,1067 @@
+    },
+    "Azure.Core.Foundations.InnerError": {
+      "type": "object",
+      "description": "An object containing more specific information about the error. As per Microsoft One API guidelines - https://github.com/Microsoft/api-guidelines/blob/vNext/Guidelines.md#7102-error-condition-responses.",
```
discussed offline, wont fix as this is auto-generated by typeSpec built-in models

raych1 commented:
@chunyu3 , can you have a look at build error for the generated SDK?
```
Start to build sdk: /mnt/vss/_work/1/s/azure-sdk-for-net/sdk/communication/Azure.Communication.Messages
05:03:40.690 cmdout 	[Invoke-GenerateAndBuildV2.ps1] MSBuild version 17.4.0+18d5aef85 for .NET
05:03:40.694 cmdout 	[Invoke-GenerateAndBuildV2.ps1] MSBUILD : error MSB1003: Specify a project or solution file. The current working directory does not contain a project or solution file.
05:03:40.799 cmderr W 	[Invoke-GenerateAndBuildV2.ps1] GeneratePackage: /mnt/vss/_work/1/s/azure-sdk-for-net/eng/scripts/Invoke-GenerateAndBuildV2.ps1:131
05:03:40.799 cmderr W 	[Invoke-GenerateAndBuildV2.ps1] Line |
05:03:40.799 cmderr W 	[Invoke-GenerateAndBuildV2.ps1]  131 |                GeneratePackage `
05:03:40.799 cmderr W 	[Invoke-GenerateAndBuildV2.ps1]      |                ~~~~~~~~~~~~~~~~~
05:03:40.800 cmderr W 	[Invoke-GenerateAndBuildV2.ps1]      | Failed to build sdk. exit code: False

```

tjprescott commented:
@glorialimicrosoft what changes were made after this was approved? Is this ready to merge?

glorialimicrosoft commented:
> @glorialimicrosoft what changes were made after this was approved? Is this ready to merge?

@tjprescott Hello Travis! There was a minor change to change channel id for ListTemplates from string to uuid. It was already a uuid for public preview, we just missed it when transforming to TypeSpec.
We still have a couple of customization for Java and JS waiting to be merged into this branch, but no other changes should be expected on the TypeSpec API itself. We are planning on merging this PR next week. Thank you!

arifibrahim4 committed:
Java Customization for Communication Message Typespec (#27919)

* Java Customization
